controllers/machinehealthcheck_controller_test.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package controllers
+
+import (
+	"context"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("MachineHealthCheck Reconciler", func() {
+	namespace := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "mhc-test"}}
+	testCluster := &clusterv1.Cluster{ObjectMeta: metav1.ObjectMeta{Namespace: namespace.Name, Name: "test-cluster"}}
+
+	BeforeEach(func() {
+		By("Ensuring the namespace exists")
+		if err := k8sClient.Get(ctx, client.ObjectKey{Name: namespace.GetName()}, namespace.DeepCopy()); apierrors.IsNotFound(err) {
+			Expect(k8sClient.Create(ctx, namespace.DeepCopy())).To(Succeed())
+		}
+		By("Creating the Cluster")
+		Expect(k8sClient.Create(ctx, testCluster.DeepCopy())).To(Succeed())
+	})
+
+	AfterEach(func() {
+		By("Deleting any MachineHealthChecks")
+		Expect(cleanupTestMachineHealthChecks(ctx, k8sClient)).To(Succeed())
+		By("Deleting the Cluster")
+		Expect(k8sClient.Delete(ctx, testCluster.DeepCopy())).To(Succeed())
```

Comment:
Why all the DeepCopy() calls?

Reply:
I was trying to avoid modifying the `namespace` and `testCluster` variables so they would be clean for each test, and they would be populated correctly when Ginkgo is building it's test graph, but I've managed to refactor to avoid the deepcopy's